File: AgentforceSDK-ReactNative-Bridge/android/src/main/java/com/salesforce/android/reactagentforce/EmployeeAgentAuthBridge.kt
/*
 * Copyright (c) 2024-present, salesforce.com, inc. All rights reserved.
 *
 * React Native bridge for Employee Agent auth via Salesforce Mobile SDK.
 * When Mobile SDK is present and initialized, isAuthSupported is true and methods delegate to the SDK.
 * Bootconfig (res/values/bootconfig.xml) supplies OAuth client id, redirect URI, scopes.
 */

package com.salesforce.android.reactagentforce

import android.util.Log
import com.facebook.react.bridge.Promise
import com.facebook.react.bridge.ReactApplicationContext
import com.facebook.react.bridge.ReactContextBaseJavaModule
import com.facebook.react.bridge.ReactMethod
import com.facebook.react.bridge.WritableNativeMap
import com.salesforce.androidsdk.app.SalesforceSDKManager
import com.salesforce.androidsdk.rest.ClientManager
import com.salesforce.androidsdk.rest.RestClient

class EmployeeAgentAuthBridge(reactContext: ReactApplicationContext) :
    ReactContextBaseJavaModule(reactContext) {

    companion object {
        private const val TAG = "EmployeeAgentAuthBridge"
        // Default account type for Salesforce Mobile SDK (matches res/values/strings.xml in SDK)
        private const val DEFAULT_ACCOUNT_TYPE = "com.salesforce.androidsdk"
    }

    override fun getName(): String = "EmployeeAgentAuthBridge"

    private fun clientManager(): ClientManager? {
        val manager = SalesforceSDKManager.getInstance() ?: return null
        return try {
            ClientManager(reactApplicationContext, DEFAULT_ACCOUNT_TYPE, true)
        } catch (e: Exception) {
            Log.e(TAG, "ClientManager init failed", e)
            null
        }
    }

    private fun credentialsFromRestClient(client: RestClient): WritableNativeMap {
        val map = WritableNativeMap()
        val info = client.clientInfo
        map.putString("instanceUrl", info?.instanceUrl?.toString() ?: "")
        map.putString("organizationId", info?.orgId ?: "")
        map.putString("userId", info?.userId ?: "")
        map.putString("accessToken", client.authToken ?: "")
        return map
    }

    @ReactMethod
    fun isAuthSupported(promise: Promise) {
        try {
            // Verify SDK is actually initialized, not just on the classpath
            val sdkManager = SalesforceSDKManager.getInstance()
            promise.resolve(sdkManager != null)
        } catch (e: Exception) {
            Log.w(TAG, "isAuthSupported: SDK not available", e)
            promise.resolve(false)
        }
    }

    @ReactMethod
    fun getAuthCredentials(promise: Promise) {
        try {
            val mgr = clientManager() ?: run {
                promise.resolve(null)
                return
            }
            val client = mgr.peekRestClient()
            if (client != null && client.authToken != null) {
                promise.resolve(credentialsFromRestClient(client))
            } else {
                promise.resolve(null)
            }
        } catch (e: Exception) {
            Log.e(TAG, "getAuthCredentials failed", e)
            promise.reject("ERROR", e.message)
        }
    }

    /**
     * Ask the Mobile SDK to refresh the current session and return updated credentials.
     * Use this when the Agentforce backend returns 401 (e.g. expired token). getRestClient()
     * causes the SDK to provide a RestClient, which may trigger an automatic token refresh.
     */
    @ReactMethod
    fun refreshAuthCredentials(promise: Promise) {
        try {
            val activity = currentActivity
            if (activity == null) {
                promise.reject("NO_ACTIVITY", "No current activity for refresh")
                return
            }
            val mgr = clientManager()
            if (mgr == null) {
                promise.reject("NOT_AVAILABLE", "Salesforce SDK not initialized")
                return
            }
            mgr.getRestClient(activity, object : ClientManager.RestClientCallback {
                override fun authenticatedRestClient(client: RestClient?) {
                    if (client != null && client.authToken != null) {
                        promise.resolve(credentialsFromRestClient(client))
                    } else {
                        promise.reject("REFRESH_FAILED", "No credentials after refresh")
                    }
                }
            })
        } catch (e: Exception) {
            Log.e(TAG, "refreshAuthCredentials failed", e)
            promise.reject("ERROR", e.message)
        }
    }

    @ReactMethod
    fun login(promise: Promise) {
        try {
            val activity = currentActivity
            if (activity == null) {
                promise.reject("NO_ACTIVITY", "No current activity for login")
                return
            }
            val mgr = clientManager()
            if (mgr == null) {
                promise.reject("NOT_AVAILABLE", "Salesforce SDK not initialized")
                return
            }
            // getRestClient kicks off login flow if no account; callback runs when RestClient is ready
            mgr.getRestClient(activity, object : ClientManager.RestClientCallback {
                override fun authenticatedRestClient(client: RestClient?) {
                    if (client != null && client.authToken != null) {
                        promise.resolve(credentialsFromRestClient(client))
                    } else {
                        promise.reject("LOGIN_FAILED", "No credentials after login")
                    }
                }
            })
        } catch (e: Exception) {
            Log.e(TAG, "login failed", e)
            promise.reject("ERROR", e.message)
        }
    }

    @ReactMethod
    fun logout(promise: Promise) {
        try {
            SalesforceSDKManager.getInstance()?.logout(currentActivity)
            promise.resolve(null)
        } catch (e: Exception) {
            Log.e(TAG, "logout failed", e)
            promise.reject("ERROR", e.message)
        }
    }
}

File: AgentforceSDK-ReactNative-Bridge/ios/Agentforce/EmployeeAgentAuthBridge.m
/*
 * Copyright (c) 2024-present, salesforce.com, inc. All rights reserved.
 *
 * React Native bridge for Employee Agent auth via Salesforce Mobile SDK.
 * When Mobile SDK is present (pod 'ReactNativeAgentforce/WithMobileSDK'), isAuthSupported is YES
 * and methods delegate to the SDK. Otherwise a stub is compiled and isAuthSupported is NO.
 */

#import <React/RCTBridgeModule.h>
#import <os/log.h>

#if __has_include(<SalesforceSDKCore/SFUserAccountManager.h>)
#import <SalesforceSDKCore/SFUserAccountManager.h>
#import <SalesforceSDKCore/SFOAuthCredentials.h>
#import <SalesforceSDKCore/SalesforceSDKManager.h>
#define HAS_SALESFORCE_SDK 1
#endif

@interface EmployeeAgentAuthBridge : NSObject <RCTBridgeModule>
@end

@implementation EmployeeAgentAuthBridge

RCT_EXPORT_MODULE(EmployeeAgentAuthBridge);

+ (BOOL)requiresMainQueueSetup {
  return NO;
}

#if HAS_SALESFORCE_SDK

/// Log the full OAuth error to the unified system log. The visible failures
/// ("invalid client credentials" / `invalid_client_id`) are connected-app/bootconfig
/// mismatches; NSError.userInfo carries the exact server error + description that the
/// generic localizedDescription drops. Visible in Console.app / Xcode under category
/// "bridge-diagnostics".
- (void)logAuthError:(NSString *)phase error:(NSError *)error {
  os_log_t log = os_log_create("com.salesforce.reactagentforce", "bridge-diagnostics");
  os_log_error(log, "[EmployeeAgentAuthBridge] %{public}@ FAILED → domain: %{public}@, code: %ld, desc: %{public}@, userInfo: %{public}@",
               phase,
               error.domain ?: @"<none>",
               (long)error.code,
               error.localizedDescription ?: @"<none>",
               error.userInfo ?: @{});
}

/// Build a richer rejection message that includes the underlying OAuth error so the
/// JS catch handler (and the customer's logs) sees the real cause, not just a generic line.
- (NSString *)detailedAuthMessage:(NSString *)fallback error:(NSError *)error {
  NSString *desc = error.localizedDescription ?: fallback;
  return [NSString stringWithFormat:@"%@ (domain=%@ code=%ld)", desc, error.domain ?: @"?", (long)error.code];
}

- (NSDictionary *)credentialsDictionaryFromUserAccount:(SFUserAccount *)userAccount {
  SFOAuthCredentials *creds = userAccount.credentials;
  if (!creds || !creds.accessToken) {
    return nil;
  }
  NSString *instanceUrl = creds.instanceUrl.absoluteString ?: @"";
  NSString *loginUrl = [NSString stringWithFormat:@"%@://%@", creds.protocol ?: @"https", creds.domain ?: @""];
  return @{
    @"instanceUrl": instanceUrl,
    @"organizationId": creds.organizationId ?: @"",
    @"userId": creds.userId ?: @"",
    @"accessToken": creds.accessToken ?: @"",
    @"refreshToken": (creds.refreshToken ?: [NSNull null]),
    @"loginUrl": loginUrl,
  };
}

RCT_EXPORT_METHOD(isAuthSupported:(RCTPromiseResolveBlock)resolve
                  rejecter:(RCTPromiseRejectBlock)reject) {
  resolve(@YES);
}

RCT_EXPORT_METHOD(getAuthCredentials:(RCTPromiseResolveBlock)resolve
                  rejecter:(RCTPromiseRejectBlock)reject) {
  SFUserAccount *currentUser = [SFUserAccountManager sharedInstance].currentUser;
  NSDictionary *creds = [self credentialsDictionaryFromUserAccount:currentUser];
  resolve(creds ?: [NSNull null]);
}

RCT_EXPORT_METHOD(refreshAuthCredentials:(RCTPromiseResolveBlock)resolve
                  rejecter:(RCTPromiseRejectBlock)reject) {
  SFUserAccount *currentUser = [SFUserAccountManager sharedInstance].currentUser;
  if (!currentUser || !currentUser.credentials.refreshToken.length) {
    reject(@"NO_SESSION", @"No logged-in user or refresh token available.", nil);
    return;
  }
  [[SFUserAccountManager sharedInstance] refreshCredentials:currentUser.credentials
                                                completion:^(SFOAuthInfo *authInfo, SFUserAccount *userAccount) {
    NSDictionary *creds = [self credentialsDictionaryFromUserAccount:userAccount];
    if (creds) {
      resolve(creds);
    } else {
      reject(@"REFRESH_FAILED", @"Could not read credentials after refresh.", nil);
    }
  }
                                                   failure:^(SFOAuthInfo *authInfo, NSError *error) {
    [self logAuthError:@"refresh" error:error];
    reject(@"REFRESH_FAILED", [self detailedAuthMessage:@"Token refresh failed." error:error], error);
  }];
}

RCT_EXPORT_METHOD(login:(RCTPromiseResolveBlock)resolve
                  rejecter:(RCTPromiseRejectBlock)reject) {
  dispatch_async(dispatch_get_main_queue(), ^{
    [[SFUserAccountManager sharedInstance] loginWithCompletion:^(SFOAuthInfo *authInfo, SFUserAccount *userAccount) {
      [SFUserAccountManager sharedInstance].currentUser = userAccount;
      NSDictionary *creds = [self credentialsDictionaryFromUserAccount:userAccount];
      if (creds) {
        resolve(creds);
      } else {
        reject(@"LOGIN_FAILED", @"Could not read credentials after login.", nil);
      }
    } failure:^(SFOAuthInfo *authInfo, NSError *error) {
      [self logAuthError:@"login" error:error];
      reject(@"LOGIN_FAILED", [self detailedAuthMessage:@"Login failed." error:error], error);
    }];
  });
}

RCT_EXPORT_METHOD(logout:(RCTPromiseResolveBlock)resolve
                  rejecter:(RCTPromiseRejectBlock)reject) {
  __block id observerRef;
  id observer = [[NSNotificationCenter defaultCenter]
      addObserverForName:kSFNotificationUserDidLogout
      object:nil
      queue:[NSOperationQueue mainQueue]
      usingBlock:^(NSNotification *note) {
        [[NSNotificationCenter defaultCenter] removeObserver:observerRef];
        resolve(nil);
      }];
  observerRef = observer;
  dispatch_async(dispatch_get_main_queue(), ^{
    [[SFUserAccountManager sharedInstance] logout];
  });
}

#else

// Stub when Salesforce Mobile SDK is not linked (Service Agent–only apps).

RCT_EXPORT_METHOD(isAuthSupported:(RCTPromiseResolveBlock)resolve
                  rejecter:(RCTPromiseRejectBlock)reject) {
  resolve(@NO);
}

RCT_EXPORT_METHOD(getAuthCredentials:(RCTPromiseResolveBlock)resolve
                  rejecter:(RCTPromiseRejectBlock)reject) {
  resolve([NSNull null]);
}

RCT_EXPORT_METHOD(refreshAuthCredentials:(RCTPromiseResolveBlock)resolve
                  rejecter:(RCTPromiseRejectBlock)reject) {
  reject(@"NOT_AVAILABLE", @"Token refresh requires Salesforce Mobile SDK.", nil);
}

RCT_EXPORT_METHOD(login:(RCTPromiseResolveBlock)resolve
                  rejecter:(RCTPromiseRejectBlock)reject) {
  reject(@"NOT_AVAILABLE", @"Employee Agent auth requires Salesforce Mobile SDK. Use pod 'ReactNativeAgentforce/WithMobileSDK' and add the Mobile SDK to your app.", nil);
}

RCT_EXPORT_METHOD(logout:(RCTPromiseResolveBlock)resolve
                  rejecter:(RCTPromiseRejectBlock)reject) {
  resolve(nil);
}

#endif

@end
